Add warning for selectors that return the entire state (#2022)
reduxjs/react-redux: docs/api/hooks.md

}
```

### Development mode checks

#### Selector result stability

In development, an extra check is conducted on the passed selector. It runs the selector an extra time with the same parameter, and warns in console if it returns a different result (based on the `equalityFn` provided).

This is important, as a selector returning a materially different result with the same parameter will cause unnecessary rerenders.

```ts
// this selector will return a new object reference whenever called
// meaning the component will rerender whenever *any* action is dispatched
const { count, user } = useSelector((state) => ({
  count: state.count,
  user: state.user,
}))
```

> If a selector result is suitably stable, or memoised, it will not return a different result and thus not cause a warning to be logged.

Reviewer: ```suggestion
If a selector result is stable or memoized it will not return a different result or cause a warning to be logged.
```

🇺🇸

> By default, this will only happen when the selector is first called. You can configure the check via context, or per `useSelector` call - either to run the check always, or never.

Reviewer: ```suggestion
By default, this will only happen when the selector is first called. You can configure the check in the Provider or at each `useSelector` call.
```

```tsx title="Global setting via context"
<Provider store={store} stabilityCheck="always">
  {children}
</Provider>
```

```tsx title="Individual hook setting"
function Component() {
  const count = useSelector(selectCount, { stabilityCheck: 'never' })
  // run once (default)
  const user = useSelector(selectUser, { stabilityCheck: 'once' })
  // ...
}
```

:::info
This check is disabled for production environments.
:::

#### No-op selector check

> In development, a check is conducted on the result returned by the selector. It warns in console if the result is the same as the parameter passed in, i.e. the root state.

Reviewer: ```suggestion
In development, a check is conducted on the result returned by the selector. It warns in the console if the result is the same as the parameter passed in, i.e. the root state.
```

> A `useSelector` call returning the entire root state is almost always a mistake, as it means the component will rerender whenever _anything_ in state changes. Selectors should select as specifically as possible.

Reviewer: Wonder if we should change `state changes` -> `in the store changes`?

> A `useSelector` call returning the entire root state is almost always a mistake, as it means the component will rerender whenever _anything_ in state changes. Selectors should select as specifically as possible.

Reviewer: ```suggestion
A `useSelector` call returning the entire root state is almost always a mistake, as it means the component will rerender whenever _anything_ in state changes. Selectors should be as granular as possible.
```

```ts no-transpile

> // this selector returns the entire state, meaning that the component will rerender unnecessarily

Reviewer: ```suggestion
// BAD: this selector returns the entire state, meaning that the component will rerender unnecessarily
```

const { count, user } = useSelector((state) => state)

> // instead, select only the state you need, calling useSelector as many times as needed

Reviewer: ```suggestion
// GOOD: instead, select only the state you need, calling useSelector as many times as needed
```

const count = useSelector((state) => state.count)
const user = useSelector((state) => state.user)
```

> By default, this will only happen when the selector is first called. You can configure the check via context, or per `useSelector` call - either to run the check always, or never.

Reviewer: ```suggestion
By default, this will only happen when the selector is first called. You can configure the check in the Provider or at each `useSelector` call.
```
